compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java

```diff
@@ -650,16 +650,23 @@ public BLangNode transform(ModulePartNode modulePart) {
         compilationUnit.name = currentCompUnitName;
         compilationUnit.setPackageID(packageID);
         Location pos = getPosition(modulePart);
+        BLangIdentifier compUnit = this.createIdentifier(pos, compilationUnit.getName());
         // Generate import declarations
         for (ImportDeclarationNode importDecl : modulePart.imports()) {
             BLangImportPackage bLangImport = (BLangImportPackage) importDecl.apply(this);
-            bLangImport.compUnit = this.createIdentifier(pos, compilationUnit.getName());
+            bLangImport.compUnit = compUnit;
             compilationUnit.addTopLevelNode(bLangImport);
         }
 
         // Generate other module-level declarations
         for (ModuleMemberDeclarationNode member : modulePart.members()) {
-            compilationUnit.addTopLevelNode((TopLevelNode) member.apply(this));
+            if (member.kind() == SyntaxKind.MODULE_XML_NAMESPACE_DECLARATION) {
+                BLangXMLNS bLangXMLNS = (BLangXMLNS) member.apply(this);
+                bLangXMLNS.compUnit = compUnit;
+                compilationUnit.addTopLevelNode(bLangXMLNS);
+            } else {
+                compilationUnit.addTopLevelNode((TopLevelNode) member.apply(this));
+            }
```

Comment:
Since calling `member.apply` and `compilationUnit.addTopLevelNode` is common, I think you can factor things out to make this more clear
```suggestion
            TopLevelNode node = (TopLevelNode) member.apply(this);
            if (member.kind() == SyntaxKind.MODULE_XML_NAMESPACE_DECLARATION) {
                // If what you are really caring about is node being BLangXMLNS node replace the condition above with a pattern matching instanceOf and get rid of the cast
                BLangXMLNS bLangXMLNS = (BLangXMLNS)node;
                bLangXMLNS.compUnit = compUnit;
            } 
            compilationUnit.addTopLevelNode(node);
```